Scope API keys to a gem

jenshenny opened the pull request:
### Background
This PR allows users to select a gem scope an API key when creating and editing. Keys with this scope can only push, yank, add/remove owners with only the selected gem. All other behaviour would remain the same.

### Data model additions
New join table `ApiKeyRubygemScope` that belong to a gem ownership and an API key. An API key can up to one `ApiKeyRubygemScope`, ownerships can have many `ApiKeyRubygemScope`.

There's also a `soft_deleted` column on ApiKeys, which will be explained more in the `What happens when an ownership is deleted?` section.

### Creating/Editing a Key with a Gem scope
If user finds themselves submitting an invalid rubygem id, an error will be displayed "Selected gem cannot be scoped to this key"

https://user-images.githubusercontent.com/42748004/161677305-ccff2bb3-b211-4627-a1bd-7e238e3e2133.mov



### CLI

https://user-images.githubusercontent.com/42748004/161677288-50109468-828a-4127-a272-4d94924a8187.mov

### What happens when an ownership is deleted?
When an ownership is deleted, it dependently destroys all associated ApiKeyRubygemScopes. In turn, in a before_destroy callback, the ApiKeyRubygemScope updates the `soft_deleted_at` associated ApiKey. Soft deleted API keys cannot be edited or used. The user can only delete it. Thought about `invalid` being the name but that can be confusing will Rails' validations


<img width="819" alt="Screen Shot 2022-04-05 at 12 12 16 AM" src="https://user-images.githubusercontent.com/42748004/161677398-ce18d406-0cd2-47c2-b189-40c58eab899c.png">


https://user-images.githubusercontent.com/42748004/161677262-2e61ce73-b188-487b-8776-d5b3276d4f5b.mov

sonalkr132 commented on app/controllers/api/base_controller.rb:
```diff
@@ -37,11 +37,18 @@ def verify_mfa_requirement
   end
 
+  def verify_api_key_gem_scope
+    return unless @api_key.rubygem && @api_key.rubygem != @rubygem
```
I want to highlight that rubygems.org sorely needs an authorization plugin (ex: pundit). A lot of tech debt has been accrued with checks like these all over the codebase.

jenshenny replied:
I felt a little guilty writing this line of code. Maybe delegating this logic to the model with something like `def authorized_for_gem?(gem)` would be an improvement?

> I want to highlight that rubygems.org sorely needs an authorization plugin (ex: pundit). A lot of tech debt has been accrued with checks like these all over the codebase.

Would it be worth creating an issue for this? I bet someone would be interested in looking into this (myself included).

sonalkr132 replied:
see https://github.com/rubygems/rubygems.org/issues/2967

sonalkr132 commented on test/functional/api/v1/deletions_controller_test.rb:
```diff
@@ -108,6 +108,29 @@ class Api::V1::DeletionsControllerTest < ActionController::TestCase
+        context "to another gem" do
+          setup do
+            ownership = create(:ownership, user: @user, rubygem: create(:rubygem, name: "test"))
+            @api_key.update(rubygem: ownership.rubygem)
```
This setup can lot more clearer if make use new API key with explicit names to reflect intention.

sonalkr132 commented on app/controllers/api/base_controller.rb:
```diff
@@ -37,11 +37,18 @@ def verify_mfa_requirement
     @api_key   = ApiKey.find_by_hashed_key(hashed_key)
-    render_unauthorized unless @api_key
+    return render_unauthorized unless @api_key
+    render_invalid_api_key unless @api_key.valid?
```
I am not sure if I understand in which case will the valid? check come into play. Please add a test for it if there is. If you have already added one, please make sure to update the wording to reflect that valid? is being checked.

jenshenny replied:
Yes, that would be the use case. I agree it's awkward to be using `valid?` in this case. I created a [method](https://github.com/rubygems/rubygems.org/blob/7eecd2094dd0700e094639bf317ef45d9427290a/app/models/api_key.rb#L38) under api keys for this authorization behaviour. Let me know if you had something else in mind.

sonalkr132 commented:
> should a join table be created between API keys and gems instead of a column for API keys?

IMO, we should create a join table. On the related note, I don't think we need rubygems has_many api_keys relation.

> just a textbox. Prone to typos

I think we should start with this. We would show validation errors for typos.

simi commented on app/controllers/api/base_controller.rb:
```diff
@@ -37,11 +37,18 @@ def verify_mfa_requirement
+  def verify_api_key_gem_scope
+    return unless @api_key.rubygem && @api_key.rubygem != @rubygem
+
+    render plain: "You do not have permission. This API key is scoped to #{@api_key.rubygem.name}.", status: :forbidden
```
Should we really report back which gem is assigned to the key? I would expect same message with wrong (non-existing key) as for key with different scope. It is just invalid for this request (and it doesn't matter why). Exposing this info can make it it easier when you find out a key without having knowledge which gem is assigned to it.

Yes, in theory you can try to find all possible gems to the key and try one by one, but what to make it easier?

jenshenny replied:
My intent was to display why the request was invalid as much as possible, but it's true that we are making it easier for someone malicious to use this key as you mentioned. I think it's sufficient to say that the key isn't scoped to this gem, and the user can view the key on the UI to figure out which gem the key belongs to 👍

simi replied:
I think "invalid key" message and "key belongs to different gem" messages should be equal.

jenshenny replied:
I see what you mean. However, the behaviour between these two cases has a slight difference. For the invalid case, the key cannot be used in **any** api request while the key belongs to a different gem cannot be used in some **owner and gem specific commands**, so I'm having a bit of trouble crafting a consistent message. 

For now, I have it as `"Rubygem <name> cannot be scoped this API key".` but that does not sounds entirely correct for either cases. Instead I was thinking of changing  in the "key belongs to a different gem" message to be something like this:

`This API key does not have permission to modify this gem`

while in the "invalid key case" have something more generic since it is set up so that there could be multiple reasons why something is invalid (eg. implementing the concept of expiries on keys).

`This API key is invalid`

wdyt?

tomstuart commented:
> > Do we want to support selection of a subset of the user’s gems instead of just one (now or in the future)? If so, should a join table be created between API keys and gems instead of a column for API keys?
>
> IMO, we should create a join table.

Can you say anything about why you’d prefer a join table? Are you saying that we _do_ want to support multiple gems per API key? If not (or not yet) then IMO it may be clearer and simpler to stick with a `belongs_to :rubygem` for now, and migrate to a join table if it becomes necessary in future.

sonalkr132 commented:
I am primarily concerned about api_table.rubygems_id pointing to gems that have changed ownership.
  - user1 creates key1 for gem1
  - user1 yanks gem1 (all versions)
  - user2 claims gem1 after 100 days lock-in
  - key1 of user1 still gives user1 access to gem1 (should not happen)
  
If we keep the association separate, we can delete the record (with ownership). I understand we have an authorization check in code (`return true if rubygem.owned_by?(user)`) that will prevent this without having to delete records, However, I much rather keep table data consistent as well. 

> Are you saying that we do want to support multiple gems per API key? If not (or not yet) 

I am no opinion about support for multiple gems support, if we receive a feature request with this, we will think about it. I do want to support limiting the scope to IP CIDR(s). Again, technically it can be done without a separate table but I would prefer keeping authorization data separate from data about api key.

tomstuart commented:
Thank you @sonalkr132, that’s a very helpful answer.

If I understand correctly, you’d like to ensure that we clean up the table data when gem ownership changes (i.e. an `Ownership` is deleted) so that we can trust the correctness of the data and aren’t completely reliant upon authorization checks in code. So, for example, we could introduce an `ApiKeyAuthorization` model which joins an API key to an ownership (rather than to a user), and say that `Ownership` `has_many :api_key_authorizations, dependent: :destroy` so that the authorizations get deleted along with the ownership. Is that roughly what you had in mind?

We need to figure out how this should interact with the lifetime of an API key itself. @jenshenny pointed out that, if a key is created to be scoped to a single gem, it would be dangerous for it to then revert to an unscoped key when the ownership (and perhaps join record) is deleted. Does that mean we should still retain some data about key scope when an ownership is deleted, or alternatively that the key itself should be deleted at the same time? If a person loses ownership of a gem and later regains it, should their old API key start working again? If the `rubygem_id` lives on the API key itself then we remember the intent to scope the key; if we delete the join record then we need some other mechanism to ensure that the key doesn’t become unscoped.

> I do want to support limiting the scope to IP CIDR(s). Again, technically it can be done without a separate table but I would prefer keeping authorization data separate from data about api key.

Okay, so you’re suggesting the IP CIDR scope could also live on the join model (e.g. `ApiKeyAuthorization` above) and be deleted when the ownership is deleted?

jenshenny commented on app/views/api_keys/new.html.erb:
```diff
@@ -15,6 +15,12 @@
+    <div class="form__group">
+      <%= label_tag :rubygem, t(".rubygem_scope"), class: "form__label" %>
+      <p><%= t(".rubygem_scope_info") %></p>
+      <%= f.text_field :rubygem_name, value: @api_key.rubygem&.name, placeholder: t(".all_gems"), class: "form__input" %>
```
> >just a textbox. Prone to typos
>
> I think we should start with this. We would show validation errors for typos.

I got a textbox working for this with a case for typos and a case where the owner doesn't own the gem specified. I put that logic in the model so I could have all the validations in one place. It's in 2 commits atm 8fbbfa68aef060db933856ddb79b6799082617cb ce0c70659e37fa8fd0afdf7a3ae40fb6b5dabce3.

 I tried having the logic in controller also (it's commented out), but I had to early return and display the typo message before other validations since calling save effectively regenerates and clears previous errors. I'm keen to hear if anyone has other ideas.

<details>
<summary>Create API key</summary>
<img width="883" alt="Screen Shot 2022-02-07 at 9 30 58 PM" src="https://user-images.githubusercontent.com/42748004/152907590-20a54598-755f-44fc-8741-a73220a90e52.png">
<img width="917" alt="Screen Shot 2022-02-07 at 9 29 59 PM" src="https://user-images.githubusercontent.com/42748004/152907600-154461f6-9ad3-41b1-a9b3-003681c27dd3.png">
</details>

sonalkr132 replied:
have you checked how pypi is handling this?

sonalkr132 replied:
sure. we can go back to dropdown.

sonalkr132 commented:
> Does that mean we should still retain some data about key scope when an ownership is deleted, or alternatively that the key itself should be deleted at the same time?

Yes, I would prefer dependent destroy but you are right it would mean that api_key reverts to unscoped key. Technically, we can use api_key.fk_id pointing to non-existent record as check, however,  I don't want to keep dangling pointers in api_key either. I think we don't want to delete the API key because it may not be as intuitive for the user when they are debugging why their key is not working (given that they don't remember that the key was scoped). I guess metadata about the key being invalid would be the way unless someone has a better idea.

> Storing the rubygem_id on the API key itself means that we automatically remember the intent to scope the key to a specific gem;

I understand. I would still prefer a more verbose schema over this.

> so you’re suggesting the IP CIDR scope could also live on the join model 

yes

jenshenny commented:
Sorry for the delay... I updated the PR to reflect the suggestions raised. Updated the PR description with the changes (1. being adding a join model between api keys and ownership and 2. falling back to a dropdown).

sonalkr132 commented:
A few suggestions on UI:
- if api_key.rubygem is nil, the api keys table gem column should show "all gems" (empty as of now)
- when key is invalid in addition to grey out, we should also explain why is it greyed out. perhaps with tooltip.
- figure out a better way to pairing auth scopes with gem scope. for example, user wants to create a key with access_webhook scope, they can still specify gem scope(which doesn't mean anything). This may be confusing. We can grey out the gem scope area until user selects scopes where gem scope is applicable. Perhaps we can add validation checks as well in model.

sonalkr132 commented:
A few suggestions on UI:
- if api_key.rubygem is nil, the api keys table gem column should show "all gems" (empty as of now)
- when key is invalid in addition to grey out, we should also explain why is it greyed out. perhaps with tooltip.
- figure out a better way to pairing auth scopes with gem scope. for example, user wants to create a key with access_webhook scope, they can still specify gem scope(which doesn't mean anything). This may be confusing. We can grey out the gem scope area until user selects scopes where gem scope is applicable. Perhaps we can add validation checks as well in model.

sonalkr132 commented on app/controllers/api_keys_controller.rb:
```diff
@@ -26,10 +27,18 @@ def create
 
   def edit
     @api_key = current_user.api_keys.find(params.require(:id))
+    return unless @api_key.soft_deleted?
```
Do we need to add error message and forbidden status here?

sonalkr132 replied:
Oh, ignore my previous comment. msg and status is needed "unless" key is soft deleted.

sonalkr132 commented on app/controllers/api_keys_controller.rb:
```diff
@@ -26,10 +27,18 @@ def create
 
   def edit
     @api_key = current_user.api_keys.find(params.require(:id))
+    return unless @api_key.soft_deleted?
```
Do we need to add error message and forbidden status here?

sonalkr132 commented on test/functional/api_keys_controller_test.rb:
```diff
@@ -124,6 +124,28 @@ class ApiKeysControllerTest < ActionController::TestCase
+
+        should "display error with invalid id" do
+          post :create, params: { api_key: { name: "gem scope", add_owner: true, rubygem_id: -1 } }
+          Delayed::Worker.new.work_off
```
why do we need work_off in this test and at 135?

sonalkr132 replied:
why do we need work_off in this test and at 135?

sonalkr132 commented on app/models/api_key.rb:
```diff
@@ -29,6 +34,18 @@ def mfa_enabled?
   end
 
+  def rubygem_id=(id)
+    self.ownership = id.blank? ? nil : user.ownerships.find_by!(rubygem_id: id)
```
It is possible to handle `ActiveRecord::RecordNotFound` raised here with validation error message?

jenshenny replied:
Yes, however when validations are called for save in the api key controller, the current validations will be cleared. So we would need to check for errors beforehand (https://github.com/rubygems/rubygems.org/pull/2944/commits/3f7168f4a527adfafcb500a273d3d36b30e3b349). Let me know if you prefer this rather than rescuing.

sonalkr132 commented on app/controllers/api_keys_controller.rb:
```diff
@@ -41,6 +50,9 @@ def update
       flash[:error] = @api_key.errors.full_messages.to_sentence
       render :edit
     end
+  rescue ActiveRecord::RecordNotFound
```
Is there a test for this?

jenshenny replied:
Yes, should be testing it [here](https://github.com/Shopify/rubygems.org/blob/3f7168f4a527adfafcb500a273d3d36b30e3b349/test/functional/api_keys_controller_test.rb#L226).

sonalkr132 commented on app/controllers/api/base_controller.rb:
```diff
@@ -41,7 +47,8 @@ def authenticate_with_api_key
     @api_key   = ApiKey.find_by_hashed_key(hashed_key)
-    render_unauthorized unless @api_key
+    return render_unauthorized unless @api_key
+    render_soft_deleted_api_key if @api_key.soft_deleted?
```
I am not sure if we are testing rest of verify with otp only once, however soft delete should be tested tested for all actions where it impacts.

jenshenny replied:
Done via https://github.com/rubygems/rubygems.org/pull/2944/commits/dae1b58f18962819efd2c8f930db572970da7633

jenshenny commented:
I'm not much of a UX expert, but I've done what I can to make the UX better as suggested
- [x] if api_key.rubygem is nil, the api keys table gem column should show "all gems" (empty as of now) (https://github.com/rubygems/rubygems.org/pull/2944/commits/50aa4bef2b7da36636e379a1310e24a941960b07)
- [x] when key is invalid in addition to grey out, we should also explain why is it greyed out. perhaps with tooltip. (https://github.com/rubygems/rubygems.org/pull/2944/commits/803437f96c2c6f89e9042a34cd8720146c7f2a6e). I also decided to save the removed rubygem name to provide more information (https://github.com/rubygems/rubygems.org/pull/2944/commits/228b7221ec6cd66459d21be1ca0b754e8f739a29).
- [x] figure out a better way to pairing auth scopes with gem scope. for example, user wants to create a key with access_webhook scope, they can still specify gem scope(which doesn't mean anything). This may be confusing. We can grey out the gem scope area until user selects scopes where gem scope is applicable. Perhaps we can add validation checks as well in model. (https://github.com/rubygems/rubygems.org/pull/2944/commits/33d2f3bd5a23e5b5e0181ed49974dbfcda86f3c8 for grey out selector, https://github.com/rubygems/rubygems.org/pull/2944/commits/b9c7476552d1286973fda28593c7fec3a766c77a for api key validations)

<img width="1072" alt="Screen Shot 2022-05-06 at 10 41 56 PM" src="https://user-images.githubusercontent.com/42748004/167206487-de093526-8a88-474b-a55e-ace0fa89095d.png">

https://user-images.githubusercontent.com/42748004/167208203-48298f91-17a5-442a-a587-2b7e78e8f549.mov

sonalkr132 commented:
Thank you so much for working on this 💯 You can reach out in rubygems-org slack if you would like to test this in staging.
